core/orchestrator.go

```diff
@@ -856,16 +862,23 @@ func (rtm *RemoteTranscoderManager) selectTranscoder() *RemoteTranscoder {
 	}
 
 	for checkTranscoders(rtm) {
-		last := len(rtm.remoteTranscoders) - 1
-		currentTranscoder := rtm.remoteTranscoders[last]
-		if _, ok := rtm.liveTranscoders[currentTranscoder.stream]; !ok {
-			// transcoder does not exist in table; remove and retry
-			rtm.remoteTranscoders = rtm.remoteTranscoders[:last]
-			continue
-		}
-		if currentTranscoder.load == currentTranscoder.capacity {
-			// Head of queue is at capacity, so the rest must be too. Exit early
-			return nil
+		currentTranscoder, transcoderExists := rtm.streamSessions[sessionId]
+		// Assign new transcoder if not assigned, or at full load
+		if !transcoderExists || currentTranscoder.load == currentTranscoder.capacity {
```

Comment:
While this is logically correct we can avoid this OR statement by splitting the IF statement.

Also WDYT about rewording `transcoderExists` to something like `sessionExists` to avoid confusion with the comment on L8714where we check whether a transcoder has an open gRPC stream with the Orchestrator.

```go
if sessionExists {
  return currentTranscoder
}

last := len(rtm.remoteTranscoders) - 1
// ... rest of the logic
```

Reply:
Missed something in my earlier comment, my bad. 

If we do 

```
if sessionExists { return currentTranscoder }
```

We need to remove the currentTranscoder from the `streamSessions` if its connection terminates for some reason before hitting this code path. Otherwise we will try to transcode against a closed connection which will error out and only then select a new transcoder. 

So we also want to implement https://github.com/livepeer/go-livepeer/pull/1849#issuecomment-824731915 and https://github.com/livepeer/go-livepeer/pull/1849#issuecomment-824737313

Reply:
If I understand right, we would need to increment load and re-sort the transcoders even if we are using a transcoder that had been previously assigned. So I guess we would have to repeat those statements (which are currently outside the condition) if we do the early return. Please correct me incase I am wrong here, how do you think we should proceed?

Reply:
Currently load is incremented and decremented whenever a segment completes (regardless from which stream it is). 

In this PR we are changing this behaviour to only increment when a transcoder is assigned a stream (tracked by sessionId in the sessionsMap) so we only have to decrement load when it gets un-assigned.

(The added bonus of that is that -maxSessions behaviour is now the same for the O and the T) 

We can do this easily by repurposing `completeTranscoder` in this comment https://github.com/livepeer/go-livepeer/pull/1849#discussion_r618294959